linenoise/linenoise/linenoise.h

```diff
@@ -0,0 +1,335 @@
+ * @brief Stores a list of completion strings.
+ */
+typedef struct linenoise_completions {
+  size_t len;   /**< Number of completions. */
+  char **cvec;  /**< Array of completion strings. */
+} linenoise_completions_t;
+
+/**
+ * @brief Callback function for providing completions.
+ *
+ * @param str Input string from the user.
+ * @param lc Completions object to be populated.
+ */
+typedef void (*linenoise_completion_callback)(const char *str, linenoise_completions_t *lc);
+
+/**
+ * @brief Callback function for providing hints during input.
+ *
+ * @param str Input string.
+ * @param color Pointer to set the color of the hint text.
+ * @param bold Pointer to set the bold attribute (non-zero for bold).
+ * @return A dynamically allocated hint string or NULL if no hint is available.
+ */
+typedef char* (*linenoise_hints_callback)(const char *str, int *color, int *bold);
+
+/**
+ * @brief Callback function to free memory returned by hints callback.
+ *
+ * @param ptr Pointer to the hint string to be freed.
+ */
+typedef void (*linenoise_free_hints_callback)(void *ptr);
+
+/**
+ * @brief Function pointer type for reading bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer to store the read bytes.
+ * @param count Number of bytes to read.
+ * @return Number of bytes read, or -1 on error.
+ */
+typedef ssize_t (*linenoise_read_bytes_fn)(int fd, void* buf, size_t count);
+
+/**
+ * @brief Function pointer type for writing bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer containing bytes to write.
+ * @param count Number of bytes to write.
+ * @return Number of bytes written, or -1 on error.
+ */
+typedef ssize_t (*linenoise_write_bytes_fn)(int fd, const void* buf, size_t count);
+
+/**
+ * @brief Structure defining the parameters needed by a linenoise
+ * instance
+ */
+typedef struct linenoise_instance_param {
+    const char *prompt;
+    size_t max_cmd_line_length;
+    int history_max_length;
+    int in_fd;
+    int out_fd;
+    bool allow_multi_line;
+    bool allow_empty_line;
+    bool allow_dumb_mode;
+    linenoise_completion_callback completion_cb;
+    linenoise_hints_callback hints_cb;
+    linenoise_free_hints_callback free_hints_cb;
+    linenoise_read_bytes_fn read_bytes_fn;
+    linenoise_write_bytes_fn write_bytes_fn;
+    char **history;
+} linenoise_instance_param_t;
+
+/**
+ * @brief Opaque handle to a linenoise instance.
+ */
+typedef struct linenoise_instance *linenoise_handle_t;
+
+/**
+ * @brief Returns the default parameters for creating a linenoise instance.
+ *
+ * @return A populated linenoise_instance_param_t structure with default values.
+ */
+linenoise_instance_param_t linenoise_get_instance_param_default(void);
+
+/**
+ * @brief Creates a new linenoise instance.
+ *
+ * @param param Pointer to the configuration parameters for the instance.
+ * @return Handle to the created instance, or NULL on failure.
+ */
+linenoise_handle_t linenoise_create_instance(linenoise_instance_param_t *param);
```

Comment:
```suggestion
esp_err_t linenoise_create_instance(const linenoise_instance_param_t *param, linenoise_handle_t* out_handle);
```

1. Configuration structure can be a const pointer
2. You may need to return different types of errors (out of RAM, invalid argument in the config structure, etc), which simply returning NULL doesn't allow. Hence, return esp_err_t and get the instance handle as an output argument.